src/optparse.coffee

```
@@ -75,41 +67,98 @@ exports.OptionParser = class OptionParser
-  args = args[..]
-  result = []
-  for arg in args
-    if match = arg.match MULTI_FLAG
-      result.push '-' + l for l in match[1].split ''
+normalizeArguments = (args, flagDict) ->
+  rules = []
+  positional = []
+  needsArgOpt = null
+  for arg, argIndex in args
+    # If the previous argument given to the script was an option that uses the
+    # next command-line argument as its argument, create copy of the option's
+    # rule with an `argument` field.
+    if needsArgOpt?
+      # FIXME: use object spread when that gets merged (in #4493)
+      rules.push Object.assign({}, needsArgOpt.rule, {argument: arg})
+      needsArgOpt = null
+      continue
+
+    multiFlags = arg.match(MULTI_FLAG)?[1]
+      .split('')
+      .map (flagName) -> "-#{flagName}"
+    if multiFlags?
+      multiOpts = multiFlags.map (flag) ->
+        rule = flagDict[flag]
+        unless rule?
+          throw new Error "unrecognized option #{flag} in multi-flag #{arg}"
+        {rule, flag}
+      # Only the last flag in a multi-flag may have an argument.
+      [innerOpts..., lastOpt] = multiOpts
+      for {rule, flag} in innerOpts
+        if rule.hasArgument
+          throw new Error "cannot use option #{flag} in multi-flag #{arg} except
+          as the last option, because it needs an argument"
+        rules.push rule
+      if lastOpt.rule.hasArgument
+        needsArgOpt = lastOpt
+      else
+        rules.push lastOpt.rule
+    else if ([LONG_FLAG, SHORT_FLAG].some (pat) -> arg.match(pat)?)
+      # TODO: do we need to check if `arg` matches regex if we already check if
+      # it's in flagDict?
+      singleRule = flagDict[arg]
+      unless singleRule?
+        throw new Error "unrecognized option #{arg}"
+      if singleRule.hasArgument
+        needsArgOpt = {rule: singleRule, flag: arg}
+      else
+        rules.push singleRule
     else
-      result.push arg
-  result
+      # This is a positional argument.
+      # TODO: should we check these with `isCoffee`?
```

In a long ago version of this PR I checked positional arguments to see whether they were recognized coffeescript files (`.coffee`, `.litcoffee`) or `--`. If an argument is provided that is not such a script, it looks like `compilePath` in `command.coffee` will just silently return, which seems unhelpful. An alternative would be to show a warning if a positional argument is given which `compilePath` won't do anything to.